platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/merkle/queue/QueueNode.java

```diff
@@ -105,59 +156,143 @@ public long getClassId() {
+        final OnDiskValue<E> value = getStore().get(new SingleLongKey(index));
+        if (value == null) {
+            throw new IllegalStateException("Can't find queue element at index " + index + " in the store");
+        }
+        return value;
     }
 
     /** Iterate over all elements */
     public Iterator<E> iterator() {
+        final QueueNodeState state = getState();
+        final QueueIterator it = new QueueIterator(state.getHead(), state.getTail());
         // Log to transaction state log, what was iterated
-        logQueueIterate(getLabel(), getRight());
-        final var itr = getQueue().stream().iterator();
-        return new Iterator<>() {
-            @Override
-            public boolean hasNext() {
-                return itr.hasNext();
-            }
+        logQueueIterate(getLabel(), state.getTail() - state.getHead(), it);
+        it.reset();
+        return it;
+    }
 
-            @Override
-            public E next() {
-                final var valueLeaf = itr.next();
-                return valueLeaf == null ? null : valueLeaf.getValue();
-            }
-        };
+    /** Utility shorthand to get the state child node */
+    @NonNull
+    private QueueNodeState getState() {
+        return getLeft();
     }
 
-    /** Utility shorthand to get the queue */
-    private FCQueue<ValueLeaf<E>> getQueue() {
+    /** Utility shorthand to get the store child node */
+    @NonNull
+    private VirtualMap<SingleLongKey, OnDiskValue<E>> getStore() {
         return getRight();
     }
+
+    /**
+     * Creates a new virtual map to be used as a data store for this node. For now, map configuration
+     * is hard-coded to use a MerkleDb data source with 2M keys hint.
+     */
+    private VirtualMap<SingleLongKey, OnDiskValue<E>> createDataStore(final String label) {
+        final KeySerializer<SingleLongKey> keySerializer = new SingleLongKeySerializer();
+        final ValueSerializer<OnDiskValue<E>> valueSerializer =
+                new OnDiskValueSerializer<>(elementSerializerClassId, elementClassId, codec);
+        final MerkleDbTableConfig merkleDbTableConfig = new MerkleDbTableConfig((short) 1, DigestType.SHA_384);
+        merkleDbTableConfig.maxNumberOfKeys(2_000_000);
+        final VirtualDataSourceBuilder dsBuilder = new MerkleDbDataSourceBuilder(merkleDbTableConfig);
+        return new VirtualMap<>(label, keySerializer, valueSerializer, dsBuilder);
+    }
+
+    private class QueueIterator implements Iterator<E> {
+
+        private final long start;
+        private final long limit;
+        private long current;
+
+        // Start (inc), limit (exc)
+        public QueueIterator(final long start, final long limit) {
+            this.start = start;
+            this.limit = limit;
+            reset();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return current < limit;
+        }
+
+        @Override
+        public E next() {
+            try {
```

Comment:
nit: it makes sense to have an assert or a check throwing `NoSuchElementException` if `current == limit`

Reply:
Done